### PR TITLE
demo

### PR DESCRIPTION
**Pull request by anhminh3105**
adding a prediction demo

### CODE REVIEW

**Comment by anhminh3105** [comment]
This new feature has not been tested on multi-device multi-gpu context.

Due to personal preference, I have changed the misc/ dir from tmp/ and didn't reflect the change for the default value of param `_C.PREDICT.LABEL_FILE_PATH = './tmp/kinetics_400_labels.csv'`.

File: misc/kinetics_400_labels.csv [hunk 1]
```diff
@@ -0,0 +1,401 @@
+id,name
```
Comment (takatosp1) [inline misc/kinetics_400_labels.csv]: Could we move it under demo/

Comment (anhminh3105) [reply misc/kinetics_400_labels.csv]: i will change that and the default config value as well

File: configs/Kinetics/demo/SLOWFAST_8x8_R50.yaml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  NUM_CLASSES: 400
+  ARCH: slowfast
+  LOSS_FUNC: cross_entropy
+  DROPOUT_RATE: 0.5
+TEST:
+  ENABLE: False
+  DATASET: kinetics
+  BATCH_SIZE: 64
+  CHECKPOINT_TYPE: caffe2
+  CHECKPOINT_FILE_PATH: "./checkpoints/SLOWFAST_8x8_R50.pkl"
```
Comment (takatosp1) [inline configs/Kinetics/demo/SLOWFAST_8x8_R50.yaml]: Might be good to give user freedom to fill the path to the checkpoint themself. Could we have some comments like:
`
# path to the checkpoint or path to the pretrain model

File: misc/kinetics_400_labels.csv (hunk 1 above)
Comment (takatosp1) [inline misc/kinetics_400_labels.csv]: Could we put this under the demo folder?

File: slowfast/config/defaults.py [hunk 3]
```diff
@@ -92,6 +93,17 @@
...
 
 
+# ---------------------------------------------------------------------------- #
+# Predicting options
+# ---------------------------------------------------------------------------- #
+_C.PREDICT = CfgNode()
+
+_C.PREDICT.ENABLE = True
+
+_C.PREDICT.LABEL_FILE_PATH = './tmp/kinetics_400_labels.csv'
```
Comment (takatosp1) [inline slowfast/config/defaults.py]: Maybe it is better to keep the default to be `""` to avoid breaking other use cases.
`_C.PREDICT.LABEL_FILE_PATH =  ""`

File: slowfast/config/defaults.py [hunk 4]
```diff
@@ -92,6 +93,17 @@
...
+
+
+
```
Comment (takatosp1) [inline slowfast/config/defaults.py]: Would be good to run the linter :P
`bash linter.sh`
It will help to fix some of the lint issues we have in this PR.

File: tools/predict_net.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+import cv2
+import torch
+
+import slowfast.utils.checkpoint as cu
+import slowfast.utils.distributed as du
+import slowfast.utils.logging as logging
+import slowfast.utils.misc as misc
+from slowfast.datasets import loader
+from slowfast.models import model_builder
+# from slowfast.utils.meters import TestMeter
```
Comment (takatosp1) [inline tools/predict_net.py]: Seems we do not need to keep this line :P

File: tools/predict_net.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
+from time import time
+
+import numpy as np
+import pandas as pd
+import cv2
```
Comment (takatosp1) [inline tools/predict_net.py]: In order to make this diff self-contained, do you mind if you could mention the opencv dependency in
`
INSTALL.md
setup.py
setup.cfg
`

File: tools/predict_net.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+                        interpolation=cv2.INTER_AREA)
+    new_h, new_w, _ = new_im.shape
+    padded_im = np.full((to_h, to_w, 3), 128)
+    x1 = (to_w-new_w)//2
+    x2 = x1 + new_w
+    y1 = (to_h-new_h)//2
+    y2 = y1 + new_h
+    padded_im[y1:y2, x1:x2, :] = new_im 
+    
+    return padded_im
```
Comment (takatosp1) [inline tools/predict_net.py]: Seems to me that we could follow what we have in `datasets/kinetics.py`?
Something like
```
            frames = transform.random_short_side_scale_jitter(
                frames, dsize, dsize
            )
```
I am not sure how would it perform with mean padding: normally the input is without padding, which means the model never seen a padded image. 

I feel two options might lead a better choice (performance?):

1) center crop:
```
            frames = transform.uniform_crop(frames, crop_size, spatial_idx)
```

or 
2) fully convolutional testing - ideally this would give us a better performance by feeding the frames without padding or cropping, then having a global pooling in the end of the network, which is implemented in [here](https://github.com/facebookresearch/SlowFast/blob/cddf2ea7c1b0fc4452e8269b977fc11a09ad2387/slowfast/models/head_helper.py#L90-L91):

I could also refactor this part later if you prefer to keep it.

File: tools/predict_net.py [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+        self.source = source
+        try:  # OpenCV needs int to read from webcam
+            self.source = int(source)
+        except ValueError:
+            pass
+
+    def __iter__(self):
+        self.cap = cv2.VideoCapture(self.source)
+        # self.cap.set(cv2.CAP_PROP_FRAME_WIDTH, 1280)
+        # self.cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 720)
```
Comment (takatosp1) [inline tools/predict_net.py]: Remove the comments if it is not used. Maybe adding an option to the config if you prefer.

File: tools/predict_net.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    model.eval()
+    misc.log_model_info(model)
+
+   # Load a checkpoint to test if applicable.
+    if cfg.TEST.CHECKPOINT_FILE_PATH != "":
+        ckpt = cfg.TEST.CHECKPOINT_FILE_PATH
+        convert_from_caffe2 = cfg.TEST.CHECKPOINT_TYPE == "caffe2"
+    elif cu.has_checkpoint(cfg.OUTPUT_DIR):
+        ckpt = cfg.OUTPUT_DIR
+        convert_from_caffe2 = False
```
Comment (takatosp1) [inline tools/predict_net.py]: might be good to avoid the hard code here?

File: tools/predict_net.py [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    logger.info(cfg)
+    # Build the video model and print model statistics.
+    model = model_builder.build_model(cfg)
+    model.eval()
+    misc.log_model_info(model)
+
+   # Load a checkpoint to test if applicable.
+    if cfg.TEST.CHECKPOINT_FILE_PATH != "":
+        ckpt = cfg.TEST.CHECKPOINT_FILE_PATH
+        convert_from_caffe2 = cfg.TEST.CHECKPOINT_TYPE == "caffe2"
```
Comment (takatosp1) [inline tools/predict_net.py]: How about move the `convert_from_caffe2` to line 121?

File: tools/predict_net.py [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+    elif cu.has_checkpoint(cfg.OUTPUT_DIR):
+        ckpt = cfg.OUTPUT_DIR
```
Comment (takatosp1) [inline tools/predict_net.py]: Could you share me the reason that you are not preferring of using 
`last_checkpoint = cu.get_last_checkpoint(cfg.OUTPUT_DIR)`
In `train_net.py` or `'test_net;.py`? I am happy to modify the api to make it easier to use for your use case.

File: tools/predict_net.py [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+    s = 0.
+    i = -1
+    for able_to_read, frame in img_provider:
+        i += 1
+        if not able_to_read:
+            # when reaches the end frame, clear the buffer and continue to the next one.
+            frames = []
+            continue
+
+        if len(frames) != cfg.DATA.NUM_FRAMES*cfg.DATA.SAMPLING_RATE:
```
Comment (takatosp1) [inline tools/predict_net.py]: `bash linter.sh` would help here I guess :P

File: tools/predict_net.py [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+            # Perform color normalization.
+            inputs = torch.tensor(frames).float()
+            inputs = inputs / 255.0
+            inputs = inputs - torch.tensor(cfg.DATA.MEAN)
+            inputs = inputs / torch.tensor(cfg.DATA.STD)
+            # T H W C -> C T H W.
+            inputs = inputs.permute(3, 0, 1, 2)
+            # 1 C T H W.
+            inputs = inputs[None, :, :, :, :]
+            # Sample frames for the fast pathway.
```
Comment (takatosp1) [inline tools/predict_net.py]: You can also use the `pack_pathway_output` in `slowfast/datasets/loader.py`.
I could also refactor the pack_pathway_output later. :P

File: tools/predict_net.py [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+            else:
+                inputs = inputs.cuda(non_blocking=True)
+
+            # Perform the forward pass.
+            preds = model(inputs)
+            # Gather all the predictions across all the devices to perform ensemble.
+            if cfg.NUM_GPUS > 1:
+                preds = du.all_gather(preds)[0]
+
+            logger.info(preds.shape)
```
Comment (takatosp1) [inline tools/predict_net.py]: I guess this line is for debugging :P

File: tools/predict_net.py [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+    cu.load_checkpoint(
+        ckpt,
+        model,
+        cfg.NUM_GPUS > 1,
+        None,
+        inflation=False,
+        convert_from_caffe2=convert_from_caffe2,
+    )
+
+    # TODO: Take inputs from camera proceed below, definitely use for loop.
```
Comment (takatosp1) [inline tools/predict_net.py]: Maybe addressing the comment here? :P

File: tools/predict_net.py [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+        # Display predicted label to frame
+        cv2.putText(frame, 'Action: {}'.format(label), (20, 30), 
+                    fontFace=cv2.FONT_HERSHEY_SIMPLEX,
+                    fontScale=1, color=(0, 235, 0), thickness=3)
+        # Display prediction speed to frame
+        cv2.putText(frame, 'Speed: {:.2f}s'.format(s), (20, 60), 
+                    fontFace=cv2.FONT_HERSHEY_SIMPLEX,
+                    fontScale=1, color=(0, 235, 0), thickness=3)
+        # Display the frame
+        cv2.imshow('SlowFast', frame)
```
Comment (takatosp1) [inline tools/predict_net.py]: I will move this part to a utility func in demo when we have demo for ava.

File: tools/run_net.py [hunk 17]
```diff
@@ -31,7 +32,13 @@ def parse_args():
...
         """
     parser = argparse.ArgumentParser(
-        description="Provide SlowFast video training and testing pipeline."
+        description="Provide SlowFast video training, testing, and predicting pipeline."
+    )
+    parser.add_argument(
+        '--predict_source',
+        help='the input source to perform prediction, \
+                can either be a path to a video or index of the camera',
+        default=0,
```
Comment (takatosp1) [inline tools/run_net.py]: Maybe set it to bool rather than int?
`type=bool` with `default=False'

I like the name of data source (or predict_source as you have here), but in order to make it consistent with what we have below (line 103), do you think we could use something like:
```
enable_demo
DEMO.ENABLE
```

**Comment by takatosp1** [comment]
(I'll probably impl an async loader and predictor on top to optimize the fluency (and probably can leverage on multi-gpus if any.)

**Comment by anhminh3105** [comment]
I'm glad that you found my code interesting. Also thanks a lot for pointing out those faults and giving such advisive comments. I could try to help fix the parts you've mentions in my spare time in the evening.

Br

**Comment by anhminh3105** [comment]
Would prefer to have the `demo/` dir under root dir? I can move it out.

**Comment by takatosp1** [comment]
Thanks for the effort! You could also leave some parts to me to address if you want. 
For demo I think we have a demo folder under root dir :P

File: tools/predict_net.py (hunk 7 above)
Comment (anhminh3105) [reply tools/predict_net.py]: 2) is a good idea, I'm thinking of using slowfast.datasets.cv2_transform.scale. What do you think?

File: tools/predict_net.py (hunk 11 above)
Comment (anhminh3105) [reply tools/predict_net.py]: I didn't look carefully into `cu` and hence didn't know there would be a `checkpoints/` in `cfg.OUTPUT_DIR` XP.  I'm going to change `ckpt = cfg.OUTPUT_DIR` to `ckpt = cu.get_last_checkpoint(cfg.OUTPUT_DIR)`.  It seem
 `cu.get_last_checkpoint(...)` does similar things to `cu.has_checkpoint(...)` in the beginning, if there could be a way to reuse the function or merging them together would help to keep the code clean, yet it's not that big deal.

File: tools/run_net.py (hunk 17 above)
Comment (anhminh3105) [reply tools/run_net.py]: it's also a good idea if I understand your idea of the `DEMO.ENABLE` correctly, do you mean to have a CfgNode() for it in default.py? if it's so, I'm thinking of having a common Node for demoing prediction, something like this:
```
_C.DEMO = CfgNode()

_C.DEMO.ENABLE = True

_C.DEMO.LABEL_FILE_PATH = ""

_C.DEMO.DATA_SOURCE = 0
```
Instead of:
```
_C.PREDICT = CfgNode()

_C.PREDICT.ENABLE = True

_C.PREDICT.LABEL_FILE_PATH = ""
```
This way I don't have to add predict_source in the parsing arguments in your run_net.py.

File: tools/predict_net.py (hunk 16 above)
Comment (anhminh3105) [reply tools/predict_net.py]: I made another version to enable multi-label prediction using a threshold, what do you think?

File: tools/predict_net.py (hunk 7 above)
Comment (anhminh3105) [reply tools/predict_net.py]: 2 ) is a good idea, I'm thinking of using slowfast.datasets.cv2_transform.scale. What do you think?

**Comment by anhminh3105** [comment]
refactoring the code according to your feedback to add new features for ava

File: tools/predict_net.py (hunk 14 above)
Comment (anhminh3105) [reply tools/predict_net.py]: yep :D
